### PR TITLE
IODeviceWeight= configures bfq.io.weight  too

### PR DESCRIPTION
**Pull request by Werkov**
This is a possible extension of the PR #21728. Opening separate PR since the per-device weights can't be configured when the particular device doesn't support (it's not configured) proportional control method.

### CODE REVIEW

File: src/core/cgroup.c [hunk 1]
```diff
@@ -1279,28 +1279,34 @@ assert_cc(BFQ_WEIGHT(10000) == 1000);
...
 
         /* FIXME: drop this when distro kernels properly support BFQ through "io.weight"
          * See also: https://github.com/systemd/systemd/pull/13335 and
          * https://github.com/torvalds/linux/commit/65752aef0a407e1ef17ec78a7fc31ba4e0b360f9.
          * The range is 1..1000 apparently, and the default is 100. */
-        xsprintf(buf, "%" PRIu64 "\n", BFQ_WEIGHT(weight));
-        (void) set_attribute_and_warn(u, "io", "io.bfq.weight", buf);
+        bfq_weight = BFQ_WEIGHT(weight);
+        xsprintf(buf, "%" PRIu64 "\n", bfq_weight);
+        if (!set_attribute_and_warn(u, "io", "io.bfq.weight", buf))
```
Comment (keszybz) [inline src/core/cgroup.c]: Please don't use boolean ops for int values. This should be `if (set_attribute_and_warn(…) >= 0)`.

**Review by keszybz — COMMENTED** [review]
Looks reasonable.

**Comment by Werkov** [comment]
Force pushed on top of [changes](https://github.com/systemd/systemd/pull/21728#issuecomment-1009164338).

I'm emphasizing a comment from the commit message:
> Since per-device weight of a particular device cannot use both io.cost
policy and BFQ scheduler, at least one warning is always expected with
this approach (the method that is unconfigured).

File: src/core/cgroup.c [hunk 2]
```diff
@@ -1059,11 +1059,15 @@ static void set_bfq_weight(Unit *u, const char *controller, uint64_t io_weight)
         /* Adjust to kernel range is 1..1000, the default is 100. */
         bfq_weight = BFQ_WEIGHT(io_weight);
 
-        xsprintf(buf, "%" PRIu64 "\n", bfq_weight);
+        if (has_dev)
```
Comment (yuwata) [inline src/core/cgroup.c]: Not necessary to introduce `has_dev` argument, but just use `major(dev) > 0`.

**Review by yuwata — APPROVED** [review]
LGTM. One minor suggestion.

**Comment by yuwata** [comment]
#21728 is merged. Please rebase.

File: src/test/test-cgroup-util.c [hunk 3]
```diff
@@ -426,4 +426,13 @@ TEST(cg_get_keyed_attribute) {
         }
 }
 
+TEST(bfq_weight_conversion) {
+        assert_se(BFQ_WEIGHT(1) == 1);
+        assert_se(BFQ_WEIGHT(50) == 50);
+        assert_se(BFQ_WEIGHT(100) == 100);
+        assert_se(BFQ_WEIGHT(500) == 136);
+        assert_se(BFQ_WEIGHT(5000) == 545);
+        assert_se(BFQ_WEIGHT(10000) == 1000);
```
Comment (keszybz) [inline src/test/test-cgroup-util.c]: No, please drop this commit. It replaces a **static** test by something that actually emits code, even if not much. The static test is done by the compiler, immediately and without any further cost. And if the tests are inline in cgroup.c, there is no need to export the macros.

**Review by keszybz — CHANGES_REQUESTED** [review]

File: src/test/test-cgroup-util.c (hunk 3 above)
Comment (Werkov) [reply src/test/test-cgroup-util.c]: I was aware of the tests having been evaluated at compile time.
There's no free lunch, everyone who just compiles will need to exert this work, whereas with the extraction, it's only run when the tests are actually wanted.

Perhaps the tests could be made static in `test-cgroup-util.c`?

**Comment by Werkov** [comment]
Thanks for the review. Rebased and condition changed.

File: src/test/test-cgroup-util.c (hunk 3 above)
Comment (keszybz) [reply src/test/test-cgroup-util.c]: Well, I hope that anyone who compiles this at least runs the unit tests… It's a basic check that you should always do. But even if not, I just don't think this matters: static asserts are essentially free compared to the costs of loading the code from disk and compiling and linking.

**Comment by bluca** [comment]
> This patch extends to devices what is already done for default cgroup IO
weight in commit 2dbc45a ("cgroup: Also set io.bfq.weight"). This
is possible with kernels >= v5.4 with commit 795fe54c2a82 ("bfq: Add
per-device weight"). A warning is logged if the write fails on older kernels.

Is there any sensible and usable way to detect this? People tend to complain about warnings, especially if they are repeated. If there's an error, having something like `static bool supported` or such would make it already a bit nicer

**Comment by yuwata** [comment]
> > This patch extends to devices what is already done for default cgroup IO
> > weight in commit [2dbc45a](https://github.com/systemd/systemd/commit/2dbc45aea747f25cc1c3848fded2ec0062f96bcf) ("cgroup: Also set io.bfq.weight"). This
> > is possible with kernels >= v5.4 with commit 795fe54c2a82 ("bfq: Add
> > per-device weight"). A warning is logged if the write fails on older kernels.
> 
> Is there any sensible and usable way to detect this? People tend to complain about warnings, especially if they are repeated. If there's an error, having something like `static bool supported` or such would make it already a bit nicer

What errno we get from old kernels?

**Comment by Werkov** [comment]
That should be good old `-EINVAL`. The warnings could be switched to debug level after a single warning with some limited reliability (similar to what is done with `cgroup-compat` logging).
(OTOH, this also means systems mixing [1] kernel from 2019- with systemd from 2022+. So I don't how much it's worth adding the compat code.)

**Comment by bluca** [comment]
if it's not too complicated I'd say it's worth it, 2019 is not _that_ long ago

**Comment by Werkov** [comment]
I've added two more commits to the branch with a draft of possible kernel version aware logging. I admit I don't like it, so I'm ready drop the two commits again :-)
WDYT?

**Comment by Werkov** [comment]
And one more commit stacked to address @keszybz's testing concern.

File: src/core/cgroup.c [hunk 4]
```diff
@@ -45,10 +46,15 @@
...
- * out specific attributes from us. */
-#define LOG_LEVEL_CGROUP_WRITE(r) (IN_SET(abs(r), ENOENT, EROFS, EACCES, EPERM) ? LOG_DEBUG : LOG_WARNING)
+/* Returns the log level to use when cgroup attribute writes fail.
+ * We downgrade to LOG_DEBUG when the attribute is missing or we have access problems (this is supposed to be
+ * nice to container managers and kernels which want to mask out specific attributes from us). Downgrade also
+ * on optional list of errors. */
+#define LOG_LEVEL_CGROUP_WRITE(r, ...) (IN_SET(abs(r), ENOENT, EROFS, EACCES, EPERM, ##__VA_ARGS__) ? LOG_DEBUG : LOG_WARNING)
+
+static const char *cgroup_jump_versions[_CGROUP_JUMP_MAX] = {
+        "5.4", /* 795fe54c2a82 ("bfq: Add per-device weight"). */
```
Comment (bluca) [inline src/core/cgroup.c]: We don't usually do version comparison as enterprise kernels backport all the stuff - I'd say a static bool supported = true/false would be enough? log once, then skip it forever

Comment (Werkov) [reply src/core/cgroup.c]: FTR, there's [a precedent](https://github.com/systemd/systemd/blob/682a9f2e88dc79531c13973fee50b252447381dd/src/basic/cgroup-util.c#L2254). I added the version check (not much excited about that neither) to properly handle `EINVAL`s on the newer kernels.

OTOH (when checking the current kernel now), if we assume only the systemd well-formatted input, there should be possible only `ENODEV`, `EOPNOTSUPP`, `ERANGE` (avoided by scaling), rarely `EBUSY` or `ENOMEM`. I.e. `EINVAL` could be demoted to debug in any kernel.

I'll simplify the check then.

Comment (bluca) [reply src/core/cgroup.c]: Sounds good, thanks

**Comment by poettering** [comment]
hmm, I *really* dislike the idea of that kernel version table. Please let's *not* do that. Why? 

Quite frankly, we should have never merged that BFQ supports. It's such a mess, requiring myriads of work-arounds everywhere, because it always behaves differently than the regular scheduler. Please let's not add major infra for dealing with apparently incomplete stuff like this, in particular not kernel version checks stuff which so far we always fended off.

**Comment by Werkov** [comment]
Note that I [share the sentiment](https://github.com/systemd/systemd/pull/21795#discussion_r786190667) towards kernel version checking and simpler solution should exist (not to pollute issue tracker, this should have likely been tagged as `reviewed, needs rework` before I get down to it).

**Comment by poettering** [comment]
Anyway, the kernel version infra this adds is way over the top. find a different way, let's not add something complex like that.

I don't really grok the issue here though, i.e. there are two distinct attributes? why not write both, and only complain if both fail?

Also, why doesn't the kernel report the feature set in `/sys/kernel/cgroup/features`? That's what that file is for...

(As before: bfq appears a totally incomplete mess, I am tmpted to just remove all support for it. We shouldn't support something that pretty obviously is chaos and incomplete).

**Comment by Werkov** [comment]
Force pushed another iteration (it's also rebased):
- kernel support is simplified to mere `-EINVAL` check,
- error evaluation of two-path write to avoid unnecessary warnings,
- (restored runtime tests of [non-macro BFQ_WEIGHT](https://github.com/systemd/systemd/pull/20522#pullrequestreview-868900600) by dropping a revert patch).

PTAL.

(I'm also not delighted by the state of BFQ APIs but that's what the world evolved into and doesn't seem to change. And since there are BFQ users, at some stage the mess must be harnessed. (I see no reason for not having the `io.bfq.weight` change captured in `/sys/kernel/cgroup/features` besides it was forgotten, which makes it less useful nowadays.))

File: src/core/cgroup.c [hunk 5]
```diff
@@ -1075,11 +1075,15 @@ static void set_bfq_weight(Unit *u, const char *controller, uint64_t io_weight)
...
+        if (major(dev) > 0)
+                xsprintf(buf, "%u:%u %" PRIu64 "\n", major(dev), minor(dev), bfq_weight);
+        else
+                xsprintf(buf, "%" PRIu64 "\n", bfq_weight);
 
         if (set_attribute_and_warn(u, controller, p, buf) >= 0 && io_weight != bfq_weight)
-                log_unit_debug(u, "%sIOWeight=%" PRIu64 " scaled to %s=%" PRIu64,
+                log_unit_debug(u, "%sIO%sWeight=%" PRIu64 " scaled to %s=%" PRIu64,
                                streq(controller, "blkio") ? "Block" : "",
+                               major(dev) > 0 ? "Device" : "",
```
Comment (poettering) [inline src/core/cgroup.c]: man, the expression to figure out the prop name is just mean...

File: src/core/cgroup.c [hunk 6]
```diff
@@ -1080,7 +1082,21 @@ static void set_bfq_weight(Unit *u, const char *controller, dev_t dev, uint64_t
         else
                 xsprintf(buf, "%" PRIu64 "\n", bfq_weight);
 
-        if (set_attribute_and_warn(u, controller, p, buf) >= 0 && io_weight != bfq_weight)
+        r = cg_set_attribute(controller, u->cgroup_path, p, buf);
+
+        /* FIXME: drop this when kernels prior
+         * 795fe54c2a82 ("bfq: Add per-device weight") v5.4-rc1~148^2~20
```
Comment (poettering) [inline src/core/cgroup.c]: don't mention the rc kernel. just say 5.4. The rc kernels are not interesting the moment the next rc kernel comes out, hence no need to be precise on that.

File: src/core/cgroup.c [hunk 7]
```diff
@@ -1080,7 +1082,21 @@ static void set_bfq_weight(Unit *u, const char *controller, dev_t dev, uint64_t
...
+         * are not interesting anymore. Old kernels will fail with EINVAL, while new kernels won't return
+         * EINVAL on properly formatted input by us. Treat EINVAL specially. */
+        if (r < 0 && r == -EINVAL && major(dev) > 0 && !warned) {
```
Comment (poettering) [inline src/core/cgroup.c]: you can drop the first part of the expression entirely

File: src/core/cgroup.c [hunk 8]
```diff
@@ -1080,7 +1082,21 @@ static void set_bfq_weight(Unit *u, const char *controller, dev_t dev, uint64_t
...
+               log_unit_warning(u, "Kernel version does not accept per-device setting in %s.", p);
+               warned = true;
+        }
+        if (r < 0)
+                log_unit_full_errno(u, r == -EINVAL ? LOG_DEBUG : LOG_LEVEL_CGROUP_WRITE(r),
+                                    r, "Failed to set '%s' attribute on '%s' to '%.*s': %m",
+                                    p, empty_to_root(u->cgroup_path), (int) strcspn(buf, NEWLINE), buf);
+        else if (r >= 0 && io_weight != bfq_weight)
```
Comment (poettering) [inline src/core/cgroup.c]: the first part of the expression is redundant

File: src/core/cgroup.c [hunk 9]
```diff
@@ -1088,35 +1088,48 @@ static void set_bfq_weight(Unit *u, const char *controller, dev_t dev, uint64_t
...
+        if (r1 >= 0 && r2 >= 0)
+                return;
+        if (r1 == -EOPNOTSUPP && r2 == -EOPNOTSUPP)
+                r = -EOPNOTSUPP; /* unconfigured device */
+        else if (r1 == -EOPNOTSUPP)
+                r = r2;
+        else if (r2 == -EOPNOTSUPP)
+                r = r1;
+        else /* both fail, prefer latter errno */
+                r = r2;
```
Comment (poettering) [inline src/core/cgroup.c]: this one line does the exact same as the huge expression above, no?

```c
r = r2 == -EOPNOSTUPP ? r1 : r2;
```

File: src/core/cgroup.c (hunk 8 above)
Comment (Werkov) [reply src/core/cgroup.c]: The idea is to emit the scaling message only when the (scaled) value was really applied. That justifies the first part, doesn't it?

File: src/core/cgroup.c (hunk 9 above)
Comment (Werkov) [reply src/core/cgroup.c]: Thanks, looks like correct simplification. It takes some time to wrap head around but it's also descriptive.

File: src/core/cgroup.c (hunk 5 above)
Comment (Werkov) [reply src/core/cgroup.c]: What do you mean?
It favors log message reader over code reader.

Comment (poettering) [reply src/core/cgroup.c]: maybe do:

```c
static const char* const prop_name[] = {
        "IOWeight",
        "BlockIOWeight",
        "IODeviceWeight",
        "BlockIODeviceWeight",
};

const char *pn = prop_name[2*(major(dev) > 0) + 
                           streq(controller, blkio)];
```

Still cryptic as fuck, but at least isolated and the prop names appear literally.

File: src/core/cgroup.c (hunk 8 above)
Comment (poettering) [reply src/core/cgroup.c]: your first branch aleady checks < 0, hence the else branch doesn't have to check >= 0 anymore.

File: src/core/cgroup.c (hunk 5 above)
Comment (Werkov) [reply src/core/cgroup.c]: Force pushed. Separate commit, can be squashed (or dropped).

File: src/core/cgroup.c (hunk 8 above)
Comment (Werkov) [reply src/core/cgroup.c]: The first condition is changed, so this won't apply anymore.

**Comment by keszybz** [comment]
Let's do this. The interface is so bad that no matter what we try to do, the code will need to be ugly.

**Pull request merged**